<?xml version="1.0" encoding="utf-8"  ?>
<robot name="cartpole_v0">

	<gazebo>
        <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
            <robotNamespace>/cartpole_v0</robotNamespace>
						<robotParam>/cartpole_v0_robot_description</robotParam>
        </plugin>
    </gazebo>

	<!-- * * * Link Definitions * * * -->
  	<!-- * * * Define World Link to fix the foot rail to the world * * * -->
    <link name="world"/>
	<link name="bar_link">
		<visual>
			<origin xyz="0 0 0.4" rpy="0 0 0"/>
			<geometry>
				<box size="0.05 0.05 0.8"/>
			</geometry>
	     	<material name="brown">
				<color rgba="0.9 0.8 0.6 1.0"/>
			</material>
	   </visual>
		<collision>
			<origin xyz="0 0 0.4" rpy="0 0 0"/>
			<geometry>
				<box size="0.05 0.05 0.8"/>
			</geometry>
	   </collision>
		<inertial>
		    <origin xyz="0 0 0.4" rpy="0 0 0"/>
        	<mass value="1"/>
        	<inertia ixx="0.0535416666667" ixy="0.0" ixz="0.0" iyy="0.0535416666667" iyz="0.0" izz="0.000416666666667"/>
        </inertial>
	</link>
	<gazebo reference="bar_link">
		<kp>1000.0</kp>
        <kd>1000.0</kd>
        <mu1>0.5</mu1>
        <mu2>0.5</mu2>
        <material>Gazebo/Orange</material>
    </gazebo>

	<link name="base_link">
		<visual>
			<origin xyz="0 0 0.2" rpy="0 0 0"/>
			<geometry>
				<box size="0.4 0.2 0.2"/>
			</geometry>
	     	<material name="black">
				<color rgba="0 0 0 1.0"/>
			</material>
	   </visual>
  		<collision>
			<origin xyz="0 0 0.2" rpy="0 0 0"/>
			<geometry>
				<box size="0.4 0.2 0.2"/>
			</geometry>
	   </collision>
  		<inertial>
		    <origin xyz="0 0 0.2" rpy="0 0 0"/>
        	<mass value="2.5"/>
        	<inertia ixx="0.0166666666667" ixy="0.0" ixz="0.0" iyy="0.0416666666667" iyz="0.0" izz="0.0416666666667"/>
        </inertial>
	</link>
	<gazebo reference="base_link">
        <kp>1000.0</kp>
        <kd>1000.0</kd>
        <mu1>0.5</mu1>
        <mu2>0.5</mu2>
        <material>Gazebo/Black</material>
    </gazebo>

	<link name="foot_link">
		<visual>
			<origin xyz="0 0 0.05" rpy="0 0 0"/>
			<geometry>
				<box size="6 0.025 0.1"/>
			</geometry>
	     	<material name="white">
				<color rgba="1 1 1 1.0"/>
			</material>
	   </visual>
  		<collision>
			<origin xyz="0 0 0.05" rpy="0 0 0"/>
			<geometry>
				<box size="6 0.025 0.1"/>
			</geometry>
	   </collision>
	   <inertial>
		    <origin xyz="0 0 0.05" rpy="0 0 0"/>
        	<mass value="5"/>
        	<inertia ixx="0.00442708333333" ixy="0.0" ixz="0.0" iyy="15.0041666667" iyz="0.0" izz="15.0002604167"/>
        </inertial>
	</link>
	<gazebo reference="foot_link">
        <kp>1000.0</kp>
        <kd>1000.0</kd>
        <mu1>0.5</mu1>
        <mu2>0.5</mu2>
        <material>Gazebo/White</material>
    </gazebo>

<!-- * * * Joint Definitions * * * -->
 	<joint name="cartpole_joint" type="revolute">
    	<parent link="base_link"/>
    	<child link="bar_link"/>
    	<origin xyz="0 0 0.32" rpy="0 0 0"/>
		<dynamics damping="0.0" friction="0.1"/>
       	<limit lower="-1.57" upper="1.57" effort="1" velocity="100"/>
        <axis xyz="0 1 0"/>
	</joint>
	<joint name="foot_joint" type="prismatic">
    	<parent link="foot_link"/>
    	<child link="base_link"/>
       	<limit lower="-2.0" upper="2.0" effort="2000000" velocity="100000"/>
	</joint>
	<joint name="fixed" type="fixed">
        <parent link="world"/>
        <child link="foot_link"/>
    </joint>

<!-- * * * Transmission Definitions * * * -->
	<transmission name="pole_joint_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="cartpole_joint">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="pole_jointMotor">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>

	<transmission name="foot_joint_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="foot_joint">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      </joint>
      <actuator name="foot_jointMotor">
        <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>


	<link name="camera_link"/>
  	<joint name="camera_joint" type="fixed">
		<origin xyz="0 -3 0.25" rpy="0 0 1.5707" />
		<parent link="foot_link"/>
		<child link="camera_link"/>
	</joint>
	<gazebo reference="camera_link">
		<self_collide>0</self_collide>
		<enable_wind>0</enable_wind>
		<kinematic>0</kinematic>
		<enable_gravity>false</enable_gravity>

		<sensor name="camera" type="camera">
			<camera name="camera">
				<horizontal_fov>${69.4*3.14159/180}</horizontal_fov>
				<image>
					<width>$(arg camera_width)</width>
					<height>$(arg camera_height)</height>
					<format>RGB_INT8</format>
				</image>
				<clip>
					<near>0.01</near>
					<far>100</far>
				</clip>
			</camera>
			<plugin name="camera/camera_controller" filename="libgazebo_ros_camera.so">
				<cameraName>camera</cameraName>
				<frameName>camera_optical_frame</frameName>
				<updateRate>60</updateRate>
			</plugin>
			<always_on>true</always_on>
			<update_rate>60</update_rate>
			<visualize>0</visualize>
		</sensor>
	</gazebo>

</robot>
